campaign/campaign_war_archives/campaign_base.py

```diff
@@ -0,0 +1,123 @@
+from module.base.utils import *
+from module.campaign.campaign_base import CampaignBase as CampaignBase_
+from module.exception import CampaignNameError, ScriptEnd
+from module.base.button import Button
+from module.war_archives.assets import WAR_ARCHIVES_EX_OFF, WAR_ARCHIVES_SP_OFF, WAR_ARCHIVES_CAMPAIGN_CHECK
+from module.ui.assets import WAR_ARCHIVES_CHECK
+from module.ui.page import page_main, page_archives
+from module.logger import logger
+from module.config.dictionary import dic_archives_template
+import module.config.server as server
+
+
+class CampaignBase(CampaignBase_):
+    # Helper to determine whether this is the first run
+    first_run = True
+
+    def _set_archives_view(self, type):
+        """
+        Switch to either EX or SP view in page_archives
+
+        Args:
+            type(str): ex or sp
+        """
+        for n in range(3):
+            self.device.screenshot()
+            if type in 'ex':
+                self.appear_then_click(WAR_ARCHIVES_EX_OFF)
+            elif type in 'sp':
+                self.appear_then_click(WAR_ARCHIVES_SP_OFF)
+            else:
+                raise CampaignNameError
+
+    def _get_archives_entrance(self, name):
+        """
+        Create entrance button to target archive campaign
+        using a template acquired by event folder name
+
+        TODO: Each server have different selectable campaign
+              Need something similar to commission to scroll
+              or turn page. CN/JP have more than 4 atm.
+
+        Args:
+            name(str): event folder name
+        """
+        template = dic_archives_template[name]
+
+        sim, point = template.match_result(self.device.image)
+        if sim < 0.85:
+            raise CampaignNameError
+
+        button = area_offset(area=(-12, -12, 44, 32), offset=point)
+        color = get_color(self.device.image, button)
+        entrance = Button(area=button, color=color, button=button, name=name)
+        return entrance
+
+    def _in_archives_campaign(self):
+        """
+        Check if already in archives campaign
+        """
+        for n in range(3):
+            self.device.screenshot()
+            if self.appear(WAR_ARCHIVES_CAMPAIGN_CHECK, offset=(20, 20)):
+                return True
+        return False
+
+
+    def ui_goto_archives_campaign(self, type):
+        """
+        Transition to page_archives then to the configured
+        event campaign map
+
+        Args:
+            type (str): 'ex' or 'sp'.
+        """
+        # First run, regardless of current location
+        # Start from page_main
+        if self.first_run:
+            self.first_run = False
+            self.ui_ensure(destination=page_main)
```

Comment:
No need to goto page _main
If already in a war achieve, go back to page_achieves. If not, run as normal.

Reply:
I opted in retaining the 'first_run' behavior since ui_goto_event is also called after every map completion as well. Otherwise, on the 2nd or Xth run, time will be used to going back to page_archives and getting back to where it already was and data key only gives 15 minutes.  Of course, can't do much about reward or stop triggers those are just acceptable loss of time.

But if you think otherwise, I have no problem dropping it either.

Reply:
Oh, i understand what `first_run` is.
About the reward loop, you can disable it when using data key, and check reward after the key ends.
```
backup = self.config.cover(ENABLE_REWARD=False)
# Do something
backup.recover()
```

Reply:
Good idea but having trouble where to put these lines in which files and where. If you were suggesting cover in one file and recover in another file but variable sharing would be difficult I think.

Alternatively, I was thinking overriding `handle_reward` like `triggered_stop_condition`
```
backup = ""
if self.in_archives_campaign():
    if not self.appear(CAMPAIGN_OPEN, offset=(20, 20)):
        backup = self.config.cover(ENABLE_REWARD=False)
result = super().handle_reward()
if backup != "":
    backup.recover()
return result
```
CAMPAIGN_OPEN would be a new asset of a button displayed below the data key counter, in EN says "Not open yet" and when data key used, it changes to a timer print out.

Reply:
Using CAMPAIGN_OPEN woule be better. Since you override handle_reward, no need to cover attributes in config
```
def handle_reward(self):
    if not self.appear(CAMPAIGN_OPEN, offset=(20, 20)):
        return False
    return super().handle_reward()
```

Reply:
Is it possible to make `is_in_campaign` faster?
With no screenshot taken inside the function.

Reply:
The best I can think of is an internal bool like `first_run`, `in_campaign = False` in war_archives campaign_base. This might be a way to avoid screenshot altogether. Also may not need `first_run` if using this instead and maybe more clean to use a public setter method rather than direct variable update. But I'd think still using screenshot is safest option to check whether in that page to ensure those whether those actions can be performed or not. In theory the bool should work as well, assuming only select actions will ever interact and update it. 

campaign_base.py
```
in_campaign = False
def ui_goto_archives_campaign(self, mode='ex'):
     if not self.in_campaign:
          # Change needed pages
          self.in_campaign = True
```
war_archives.py
```
def handle_reward():
     if self.campaign.in_campaign and not self.appear(CAMPAIGN_OPEN), offset=(20, 20)):
         return False
     if super().handle_reward():
         self.campaign.in_campaign = False
         return True
     return False
def triggered_stop_condition():
    if self.campaign.in_campaign:
        # Same stuff

Reply:
How about replace
```
if self.first_run or not self.in_archives_campaign():
    result = self.ui_ensure(destination=page_archives)
```
with:
```
if self.first_run or not self.appear(WAR_ARCHIVES_CAMPAIGN_CHECK, offset=(20, 20)):
    result = self.ui_ensure(destination=page_archives)
```

Reply:
Oh right, a much simpler answer. This is what the helper method is technically doing just in a looping behavior to make sure it was in that screen before moving forward. My PC can be slow sometimes so thought it was better to give it some leeway for safety.

I think I see your point though, there is already a loop in effect via `ensure_campaign_ui()` so will inherently retry again until ensured or times out. Another loop will just slow things down.

I'll change to `self.appear(WAR_ARCHIVES_CAMPAIGN_CHECK, offset=(20, 20))` from what I have `self.in_archives_campaign()` in their respective files.

Reply:
```
    def handle_reward(self):
        if not self.appear(WAR_ARCHIVES_CAMPAIGN_CLOSED, offset=(20, 20)):
            logger.info('Delaying reward check until data key has been used up')
            return False
        return super().handle_reward()
```
Encountered Device has no attribute called image which makes sense since this is used in run almost immediately.  `hasattr()` check to mitigate this?

Reply:
Check if `Device` has attribute `image`, if not, take one screenshot.

Reply:
That does the trick.

New issue, not sure how I should cut the image for that "Not yet open' button text.  The background is relatively transparent and so getting mixed results based on selected archives stage map.

[captures](https://imgur.com/a/Y9Z4NLI)

I've tried whole text and the empty space on the left since that is where the timer would be. A fraction of the text may vary in results like the 'N' or 'O' isn't that different if the appearance relies more on color. Would offset be better to use? Alternatively, maybe use OCR like how I did with the 'Data Key' before.

Reply:
You can see how alas detect ambush evade  success and ambush evade failed. They are both white letters on a semi transparent background.
However, I have another way to detect this. `Not yet open` is on the left, blank area on the right. While it's open, white letters on the right. You can calculate the maximum or average RGB on the right to do the trick.

Reply:
Would like to clarify whether this is what you had in mind, I used `button.load_color()` and used image of the blank area
`Not yet open` RGB from 3 different images
[43.34810811 50.11027027 74.52756757]
[66.45675676 61.49837838 83.05945946]
[49.90972973 49.82108108 73.26486486]

`Open` RGB from 1 image
[90.05351351  96.7572973  111.13459459]

Use the `Open` RGB average as the expectation and compare against a live image's average? If lower then considered `Not yet open` else if higher then means `Open`? Perhaps some plus/minus leeway is needed to account for margin of error?

Average can be calculated like this, taken from [here](https://stackoverflow.com/questions/12703871/benchmarking-functions-for-computing-average-rgb-value-of-images)
`tuple(button.load_color().mean(axis=0))` which I think is numpy array

Also if in any other page, should probably safeguard this validation process with the same `WAR_ARCHIVES_CAMPAIGN_CHECK` condition

Reply:
[Created Assets](https://imgur.com/a/8mciqpy)

Those are the assets I'm using for testing, surprisingly using `self.appear()` on the `N` does work for all the screen captures but of course that would mean every server would have to provide their own image while this RGB method is generic enough that no one has to make asset updates.

Reply:
![WAR_ARCHIVES_TIMER](https://user-images.githubusercontent.com/37934724/99620248-24f20580-2a60-11eb-99ae-6c5f8aa9b1f0.png)
The info tab on top right is a 60% transparent black background. So average color of background will smaller than 255 * 40% = 102.
```
if np.mean(self.image_area(WAR_ARCHIVES_TIMER)) > 102:
    # Timer running
else:
    # Not yet open
```

Reply:
How did you come up with 60% transparency? Just a rough estimate? Great idea to box the asset as close as possible to the white text for the RGB to incorporate more white emphasis, those previous assets didn't quite make good RGB values for comparisons since it included too much of the other colors.

Ran through a Winter's Crown D3 sequence, no interruptions and handle_reward executed timely based on whether the timer was there or not. I think PR is ready unless I may have missed something. Let me know if I did.

Thanks

Reply:
Just estimate by color.

This PR will be merged soon.